lib/spack/spack/spec.py

```diff
@@ -4995,53 +4995,74 @@ def __missing__(self, key):
 spec_id_re = r"\w[\w.-]*"
 
 
+_filename_re_base = r"[/\w.-]*/[/\w/-]+\.(yaml|json)[^\b]*"
+_windows_filename_prefix = r"([A-Za-z]:)*?"
+# Spec strings require posix-style paths on Windows
+# because the result is later passed to shlex.
+_windows_filename_re = _windows_filename_prefix + _filename_re_base
+
+#: Regex for literal filenames.
+filename_reg = _windows_filename_re if is_windows else _filename_re_base
+
+
 class SpecLexer(spack.parse.Lexer):
 
     """Parses tokens that make up spack specs."""
 
+    # TODO: Determine whether using >2 lexer modes would allow us to create
+    # a separate entry (likely defined the same as `spec_id_re` at first) to
+    # specifically match version strings, variant names, compiler dependency
+    # names (with '%').
     def __init__(self):
-        # Spec strings require posix-style paths on Windows
-        # because the result is later passed to shlex
-        filename_reg = (
-            r"[/\w.-]*/[/\w/-]+\.(yaml|json)[^\b]*"
-            if not is_windows
-            else r"([A-Za-z]:)*?[/\w.-]*/[/\w/-]+\.(yaml|json)[^\b]*"
-        )
         super(SpecLexer, self).__init__(
             [
                 (
-                    r"\@([\w.\-]*\s*)*(\s*\=\s*\w[\w.\-]*)?",
-                    lambda scanner, val: self.token(VER, val),
+                    "BEGIN_PHASE",
+                    [
+                        # '@': begin a Version, VersionRange, or VersionList:
+                        (r"@([\w.\-]*\s*)*(\s*\=\s*\w[\w.\-]*)?", VER),
+                        # VersionRange syntax:
+                        (re.escape(":"), COLON),
+                        # VersionList syntax:
+                        (re.escape(","), COMMA),
+                        # '^': dependency, or "AND":
+                        (re.escape("^"), DEP),
+                        # variant syntax:
+                        (re.escape("+"), ON),
+                        (re.escape("-"), OFF),
+                        (re.escape("~"), OFF),
+                        # Compiler dependency syntax:
+                        (re.escape("%"), PCT),
+                        # This is *not* used in version string parsing.
+                        (re.escape("="), EQ),
+                        # Filenames match before identifiers, so no initial filename
+                        # component is parsed as a spec (e.g., in subdir/spec.yaml)
+                        (filename_reg, FILE),
+                        # Hash match after filename. No valid filename can be a hash
+                        # (files end w/.yaml), but a hash can match a filename prefix.
+                        (re.escape("/"), HASH),
+                        # Identifiers match after filenames and hashes.
+                        (spec_id_re, ID),
+                        # Gobble up all remaining whitespace between tokens.
+                        (r"\s+", None),
+                    ],
+                    {"EQUALS_PHASE": [EQ]},
```

Comment:
This was the other important reason for this change. imo it was really unclear how the list of tuples we have now ends up triggering changes in the lexer mode, because we were relying on the implementation detail of only allowing exactly two modes.

While we do not add any new modes in this specific change, this makes it a little easier for us to extend the spec syntax later e.g. by having more complex compiler specs. I recall @alalazo was working on exactly that (how to add a new element to spec syntax for compiler specs) a while ago and I think the refactoring in this PR will make that effort easier in the future.

Reply:
I was referring to this element of the virtual provider selection changes at https://github.com/spack/spack/pull/15569#issuecomment-1262947139!